File: lib/rest/yampi/services/yampi_catalog_service.dart
import 'package:sertton/core/catalog/dtos/brand_dto.dart';
import 'package:sertton/core/catalog/dtos/category_dto.dart';
import 'package:sertton/core/catalog/dtos/collection_dto.dart';
import 'package:sertton/core/catalog/dtos/product_dto.dart';
import 'package:sertton/core/catalog/interfaces/catalog_service.dart';
import 'package:sertton/core/global/responses/pagination_response.dart';
import 'package:sertton/core/global/responses/rest_response.dart';
import 'package:sertton/rest/yampi/mappers/yampi_brand_mapper.dart';
import 'package:sertton/rest/yampi/mappers/yampi_category_mapper.dart';
import 'package:sertton/rest/yampi/mappers/yampi_collection_mapper.dart';
import 'package:sertton/rest/yampi/mappers/yampi_product_mapper.dart';
import 'package:sertton/rest/yampi/services/yampi_service.dart';

class YampiCatalogService extends YampiService implements CatalogService {
  YampiCatalogService(super.restClient, super.envDriver);

  @override
  Future<RestResponse<PaginationResponse<ProductDto>>> fetchProducts({
    int page = 1,
    String? categoryId,
    List<String> brandsIds = const [],
    String? query,
  }) async {
    final filterParams = _buildFilterParams(
      categoryId: categoryId,
      brandsIds: brandsIds,
      query: query,
    );
    final response = await super.restClient.get(
      '/catalog/products?include=skus,brand,images,texts$filterParams&page=$page',
    );
    return response.mapBody((body) {
      if (response.isFailure) return PaginationResponse();
      return YampiProductMapper.toDtoPagination(body);
    });
  }

  @override
  Future<RestResponse<ProductDto>> fetchProduct(String productId) async {
    final response = await super.restClient.get(
      '/catalog/products/$productId?include=skus,brand,images,texts',
    );
    return response.mapBody<ProductDto>((body) {
      if (response.isFailure) return null;
      return YampiProductMapper.toDto(body);
    });
  }

  @override
  Future<RestResponse<List<CategoryDto>>> fetchCategories() async {
    final response = await super.restClient.get('/catalog/categories');
    return response.mapBody<List<CategoryDto>>((body) {
      if (response.isFailure) return [];
      return YampiCategoryMapper.toDtoList(body);
    });
  }

  @override
  Future<RestResponse<List<BrandDto>>> fetchBrands() async {
    final response = await super.restClient.get('/catalog/brands');
    return response.mapBody<List<BrandDto>>((body) {
      if (response.isFailure) return [];
      return YampiBrandMapper.toDtoList(body);
    });
  }

  @override
  Future<RestResponse<List<CollectionDto>>> fetchCollections() async {
    final response = await super.restClient.get(
      '/catalog/collections?include=products',
    );
    return response.mapBody<List<CollectionDto>>((body) {
      if (response.isFailure) return [];
      return YampiCollectionMapper.toDtoList(body);
    });
  }

  @override
  Future<RestResponse<List<ProductDto>>> fetchProductsByCollection(
    String collectionId,
  ) async {
    final response = await super.restClient.get(
      '/catalog/collections/$collectionId/products?include=images,skus,brand,texts',
    );
    return response.mapBody<List<ProductDto>>((body) {
      if (response.isFailure) return [];
      return YampiProductMapper.toDtoList(body);
    });
  }

  String _buildFilterParams({
    String? categoryId,
    List<String> brandsIds = const [],
    String? query,
  }) {
    final buffer = StringBuffer();

    if (categoryId != null) {
      buffer.write('&category_id[]=$categoryId');
    }

    for (final brandId in brandsIds) {
      buffer.write('&brand_id[]=$brandId');
    }

    if (query != null && query.isNotEmpty) {
      buffer.write('&q=$query');
    }

    return buffer.toString();
  }
}

File: src/rest/yampi/services/yampi-catalog-service.ts
import type {
  FetchProductsParams,
  ICatalogService,
} from "@/core/catalog/interfaces/catalog-service"
import type { RestClient } from "@/core/shared/interfaces/rest-client"
import {
  YampiBrandMapper,
  YampiCategoryMapper,
  YampiCollectionMapper,
  YampiProductMapper,
} from "@/rest/yampi/mappers"
import type {
  YampiBrand,
  YampiCategory,
  YampiCollection,
  YampiProduct,
  YampiProductResponse,
  YampiResponse,
} from "@/rest/yampi/types"

export const YampiCatalogService = (restClient: RestClient): ICatalogService => {
  const brandMapper = YampiBrandMapper()
  const categoryMapper = YampiCategoryMapper()
  const collectionMapper = YampiCollectionMapper()
  const productMapper = YampiProductMapper()

  return {
    async fetchBrands() {
      const response = await restClient.get<YampiResponse<YampiBrand>>("/catalog/brands")
      return response.mapBody((body) => body.data.map(brandMapper.toDomain))
    },

    async fetchCategories() {
      const response = await restClient.get<YampiResponse<YampiCategory>>("/catalog/categories")
      return response.mapBody((body) => body.data.map(categoryMapper.toDomain))
    },

    async fetchCollections() {
      const response = await restClient.get<YampiResponse<YampiCollection>>("/catalog/collections")
      return response.mapBody((body) => body.data.map(collectionMapper.toDomain))
    },

    async fetchProduct(productId) {
      restClient.clearQueryParams()
      restClient.setQueryParam("include", "skus,brand,images,texts")
      const response = await restClient.get<YampiProductResponse>(`/catalog/products/${productId}`)
      return response.mapBody((body) => productMapper.toDomain(body.data))
    },

    async fetchProducts(params: FetchProductsParams = {}) {
      restClient.clearQueryParams()
      restClient.setQueryParam("include", "skus,brand,images,texts")
      restClient.setQueryParam("page", params.page ?? 1)
      restClient.setQueryParam("category_id[]", params.categoryId)
      restClient.setQueryParam("q", params.query)
      params.brandsIds?.forEach((id, index) => {
        restClient.setQueryParam(`brand_id[${index}]`, id)
      })

      const response = await restClient.get<YampiResponse<YampiProduct>>("/catalog/products")
      return response.mapBody(productMapper.toPagination)
    },

    async fetchProductsByCollection(collectionId) {
      restClient.setQueryParam("include", "skus,brand,images,texts")
      const response = await restClient.get<YampiResponse<YampiProduct>>(
        `/catalog/collections/${collectionId}/products`,
      )
      return response.mapBody((body) => productMapper.toPagination(body).items)
    },

    async fetchSimilarProducts(productId) {
      restClient.setQueryParam("include", "skus,brand,images,texts")
      const response = await restClient.get<YampiResponse<YampiProduct>>(
        `/catalog/products/${productId}/similars`,
      )
      return response.mapBody((body) => productMapper.toPagination(body).items)
    },
  }
}
